trace_decoder/src/zero_jerigon.rs

```diff
@@ -0,0 +1,456 @@
+//! Frontend for the witness format emitted by the [`0xPolygonZero/erigon`](https://github.com/0xPolygonZero/erigon)
+//! Ethereum node (a.k.a "jerigon").
+
+use std::{
+    array,
+    collections::{HashMap, HashSet},
+    iter,
+};
+
+use anyhow::{bail, ensure, Context as _};
+use either::Either;
+use mpt_trie::{
+    partial_trie::{HashedPartialTrie, PartialTrie as _},
+    trie_ops::ValOrHash,
+};
+use nunny::NonEmpty;
+use u4::U4;
+
+use crate::wire::{Instruction, SmtLeaf};
+
+pub struct Frontend {
+    pub state: HashedPartialTrie,
+    pub code: HashSet<NonEmpty<Vec<u8>>>,
+    pub storage: HashMap<ethereum_types::H256, HashedPartialTrie>,
+}
+
+pub fn frontend(instructions: impl IntoIterator<Item = Instruction>) -> anyhow::Result<Frontend> {
+    let executions = execute(instructions)?;
+    ensure!(
+        executions.len() == 1,
+        "only a single execution is supported"
+    );
+    let execution = executions.into_vec().remove(0);
+
+    let mut visitor = Visitor {
+        path: Vec::new(),
+        frontend: Frontend {
+            state: HashedPartialTrie::default(),
+            code: HashSet::new(),
+            storage: HashMap::new(),
+        },
+    };
+    visitor.visit_node(match execution {
+        Execution::Leaf(it) => Node::Leaf(it),
+        Execution::Extension(it) => Node::Extension(it),
+        Execution::Branch(it) => Node::Branch(it),
+        Execution::Empty => Node::Empty,
+    })?;
+    let Visitor { path, frontend } = visitor;
+
+    assert_eq!(Vec::<U4>::new(), path);
+
+    Ok(frontend)
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Hash {
+    raw_hash: [u8; 32],
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Value {
+    raw_value: NonEmpty<Vec<u8>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Account {
+    nonce: u64,
+    balance: ethereum_types::U256,
+    storage: Option<Box<Node>>,
+    code: Option<Either<Hash, Code>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Code {
+    code: NonEmpty<Vec<u8>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Leaf {
+    key: NonEmpty<Vec<U4>>,
+    value: Either<Value, Account>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Extension {
+    key: NonEmpty<Vec<U4>>,
+    child: Box<Node>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Branch {
+    children: [Option<Box<Node>>; 16],
+}
+
+/// An interior execution node
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum Node {
+    Hash(Hash),
+    // BUG: these are documented, but never constructed during execution
+    // Value(Value),
+    // Account(Account),
```

Comment:
I assume you'll remove them when dealing with the backend then?

Reply:
I think this should wait for #349